cmd/protoc-gen-gorums/dev/zorums_ordered_gorums.pb.go

```diff
@@ -529,48 +529,51 @@ func (s *GorumsServer) RegisterOrderingUnaryRPCHandler(handler OrderingUnaryRPCH
 // OrderingFuture asynchronously invokes a quorum call on configuration c
 // and returns a FutureResponse, which can be used to inspect the quorum call
 // reply and error when available.
-func (c *Configuration) OrderingFuture(ctx context.Context, in *Request) (*FutureResponse, error) {
+func (c *Configuration) OrderingFuture(ctx context.Context, in *Request) *FutureResponse {
 	fut := &FutureResponse{
 		NodeIDs: make([]uint32, 0, c.n),
 		c:       make(chan struct{}, 1),
 	}
-	id, expected, replyChan, err := c.orderingFutureSend(ctx, in)
-	if err != nil {
-		return nil, err
-	}
-	go func() {
-		defer close(fut.c)
-		c.orderingFutureRecv(ctx, in, id, expected, replyChan, fut)
-	}()
-	return fut, nil
-}
-
-func (c *Configuration) orderingFutureSend(ctx context.Context, in *Request) (uint64, int, chan *orderingResult, error) { // get the ID which will be used to return the correct responses for a request
+	// get the ID which will be used to return the correct responses for a request
 	msgID := c.mgr.nextMsgID()
 
 	// set up a channel to collect replies
 	replyChan := make(chan *orderingResult, c.n)
 	c.mgr.putChan(msgID, replyChan)
 
+	expected := c.n
+
+	var msg *ordering.Message
 	data, err := proto.MarshalOptions{AllowPartial: true, Deterministic: true}.Marshal(in)
 	if err != nil {
-		return 0, 0, nil, fmt.Errorf("failed to marshal message: %w", err)
+		// In case of a marshalling error, we should skip sending any messages
```

Comment:
`goto End` will actually try to send stuff...

Reply:
Okay, maybe not... But it was a bit confusing to not return here.

Reply:
I agree, but the goroutine for the `Recv` method must be started before returning.
The other ways that I can think of to achieve the same result would be to either
start the goroutine before returning in both places:

```go
if err != nil {
	// In case of a marshalling error, we should skip sending any messages
	fut.err = fmt.Errorf("failed to marshal message: %w", err)
	go c.{{$unexportMethod}}Recv(ctx, in, msgID, expected, replyChan, fut)
	return fut
}
```

or to `defer` it like this (`defer go` is not supported):

```go
defer func() { go c.{{$unexportMethod}}Recv(ctx, in, msgID, expected, replyChan, fut) }()
```

Reply:
Actually, the only thing that happens in this case, after starting the goroutine is:

```go
	defer close(fut.c)

 	if fut.err != nil {
 		return
 	}
```

Hence, we can do this instead:

```go
if err != nil {
	// In case of a marshalling error, we should skip sending any messages
	fut.err = fmt.Errorf("failed to marshal message: %w", err)
	close(fut.c)
	return fut
}
```

I think this is more clear than actually having to start the goroutine to close the channel, even though we repeat the close operation in two places then...

Another alternative would be to not initialize the `fut.c` channel, until it is needed inside the `orderingFutureRecv` goroutine.

Reply:
Nice, I didn't think of that. I will go ahead and implement it. Since the goroutine is not started in case of an error, there should be no problem with closing the channel there.